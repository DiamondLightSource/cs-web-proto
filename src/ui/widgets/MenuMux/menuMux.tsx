import React, { useContext } from "react";

import { Widget } from "../widget";
import { PVWidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import { InferWidgetProps, StringProp, MacrosProp } from "../propTypes";
import { MacroContext } from "../../../types/macros";

export interface MenuMuxProps {
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  values: { [key: string]: string };
  selected: string;
}

export const MenuMuxComponent = (props: MenuMuxProps): JSX.Element => {
  const mappedOptions = Object.entries(props.values).map(
    ([name, value]): JSX.Element => {
      return (
        <option key={value} value={value}>
          {name}
        </option>
      );
    }
  );

  return (
    <select
      value={props.selected}
      style={{ width: "100%" }}
      onChange={props.onChange}
    >
      {mappedOptions}
    </select>
  );
};

export const SmartMenuMux = (props: {
  connected: boolean;
  symbol: string;
  values: { [key: string]: string };
}): JSX.Element => {
  const macroContext = useContext(MacroContext);
  function onChange(event: React.ChangeEvent<HTMLSelectElement>): void {
    macroContext.updateMacro(props.symbol, event.currentTarget.value);
  }
  if (macroContext.macros[props.symbol] === undefined) {
    macroContext.updateMacro(props.symbol, Object.values(props.values)[0]);
  }

  return (
    <MenuMuxComponent
      onChange={onChange}
      values={props.values}
      selected={macroContext.macros[props.symbol]}
    />
  );
};

const MenuMuxProps = {
  ...PVWidgetPropType,
  symbol: StringProp,
  values: MacrosProp
};

export const MenuMux = (
  props: InferWidgetProps<typeof MenuMuxProps>
): JSX.Element => <Widget baseWidget={SmartMenuMux} {...props} />;

registerWidget(MenuMux, MenuMuxProps, "menumux");
