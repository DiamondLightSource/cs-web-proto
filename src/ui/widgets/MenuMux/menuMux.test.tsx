import React from "react";
import { MenuMuxComponent } from "./menuMux";
import { create, ReactTestRenderer } from "react-test-renderer";
import { render } from "@testing-library/react";

let snapshot: ReactTestRenderer;

const menuMux = (
  <MenuMuxComponent
    onChange={jest.fn()}
    values={{ A: "a", B: "b" }}
    selected={"a"}
  />
);

beforeEach((): void => {
  snapshot = create(menuMux);
});

describe("<MenuMux />", (): void => {
  test("it matches the snapshot", (): void => {
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("it renders both options", (): void => {
    const { getByRole } = render(menuMux);
    const select = getByRole("combobox") as HTMLSelectElement;
    expect(select.childElementCount).toEqual(2);
  });
});
