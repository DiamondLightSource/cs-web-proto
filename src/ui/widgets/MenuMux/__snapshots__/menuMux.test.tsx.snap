// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuMux /> it matches the snapshot 1`] = `
<select
  onChange={[MockFunction]}
  style={
    Object {
      "width": "100%",
    }
  }
  value="a"
>
  <option
    value="a"
  >
    A
  </option>
  <option
    value="b"
  >
    B
  </option>
</select>
`;
